Feature/add configuration metrics

jgduarte-stratio opened the pull request:
It is really useful to be able to compare existing metrics with allowed maximums. 
A new database metric with **max_connections** column has been added.
A new set of metrics with configuration values is now available. Current available values are **max_client_conn** and **max_user_connections**

jgduarte-stratio commented:
Hi @esanchezm , I hope you are doing well. 
Would you have time to take a look to this PR soon?

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -99,10 +99,21 @@ def collect(self):
                     {"type": "gauge", "column": "pool_size",           "metric": "database_pool_size",           "help": "Configured Pool Size Limit"},
                     {"type": "gauge", "column": "reserve_pool",        "metric": "database_reserve_pool_size",   "help": "Configured Reserve Limit"},
                     {"type": "gauge", "column": "current_connections", "metric": "database_current_connections", "help": "Database connection count"},
+                    {"type": "gauge", "column": "max_connections",     "metric": "database_max_connections",   "help": "Database maximum number of allowed connections"},                    
```
Please, align the `"help"` element with the one above it

```suggestion
                    {"type": "gauge", "column": "max_connections",     "metric": "database_max_connections",     "help": "Database maximum number of allowed connections"},                    
```

jgduarte-stratio replied:
Thank you for answering so fast. 
I have added a new commit solving the code issues you raised. And some extra fixes I raised would doing some integration testing.

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -150,6 +161,23 @@ def _exportMetrics(self, results, metricPrefix, metricMappings, labelMappings, e
+        for result in results:
+            for mapping in metricMappings:                
+                if (result["key"] == mapping["key"]):
+                    value = result["value"]   
```
This assignment is not needed, you can use `result["value"]` directly in the code below

```suggestion
```

jgduarte-stratio replied:
It should be ok with new commit

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -150,6 +161,23 @@ def _exportMetrics(self, results, metricPrefix, metricMappings, labelMappings, e
+                    metrics.append({
+                        "type":   mapping["type"],
+                        "name":   metricPrefix + mapping['metric'],
+                        "value":  value,                    
```
```suggestion
                        "value":  result["value"],                    
```

jgduarte-stratio replied:
It should be ok with new commit

esanchezm commented on tests/test_collector.py:
```diff
@@ -23,8 +23,13 @@ def fetchMetricsSuccessFromPgBouncer17Mock(conn, query):
+    elif query == "SHOW CONFIG":
+        return [
+            {"key": "max_client_conn","value": 500, "changeable": "yes"},
+            {"key": "max_user_connections","value": 0, "changeable": "yes"}
```
Please, fix the style of these 2 lines so it's consistent:

```python
            {"key": "max_client_conn",      "value": 500, "changeable": "yes"},
            {"key": "max_user_connections", "value": 0,   "changeable": "yes"}
```

jgduarte-stratio replied:
It should be ok now with the latest commit

esanchezm commented on tests/test_collector.py:
```diff
@@ -42,9 +47,14 @@ def fetchMetricsSuccessFromPgBouncer18Mock(conn, query):
+    elif query == "SHOW CONFIG":
+        return [
+            {"key": "max_client_conn","value": 500, "changeable": "yes"},
+            {"key": "max_user_connections","value": 0, "changeable": "yes"}
```
Same here:

```python
            {"key": "max_client_conn",      "value": 500, "changeable": "yes"},
            {"key": "max_user_connections", "value": 0,   "changeable": "yes"}
```

jgduarte-stratio replied:
fixed

jgduarte-stratio commented:
Hi @esanchezm, I added your fix request from my original branch. I released later that github already allowed me to do it from your change request. Anyway I fixed a bad type format when reading from the config table as it's reading values as strings.

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -7,7 +7,7 @@
 
 class PgbouncersMetricsCollector():
-    def __init__(self, configs: List[PgbouncerConfig]):
+    def __init__(self, configs: List[PgbouncerConfig]):        
```
Extra spaces at the end of the line

```suggestion
    def __init__(self, configs: List[PgbouncerConfig]):
```

esanchezm commented on tests/test_collector.py:
```diff
@@ -161,6 +173,32 @@ def testShouldExportDatabasesMetrics(self):
+        metrics = getMetricsByName(collector.collect(), "pgbouncer_config_max_user_connections")
+        self.assertEqual(len(metrics), 1)
+        self.assertEqual(metrics[0]["type"], "gauge")
+        self.assertEqual(metrics[0]["value"], 0)        
```
```suggestion
        self.assertEqual(metrics[0]["value"], 0)
```

esanchezm commented on tests/test_collector.py:
```diff
@@ -161,6 +173,32 @@ def testShouldExportDatabasesMetrics(self):
+        metrics = getMetricsByName(collector.collect(), "pgbouncer_config_max_client_conn")
+        self.assertEqual(len(metrics), 1)
+        self.assertEqual(metrics[0]["type"], "gauge")
+        self.assertEqual(metrics[0]["value"], 500)        
```
```suggestion
        self.assertEqual(metrics[0]["value"], 500)
```

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -150,6 +161,22 @@ def _exportMetrics(self, results, metricPrefix, metricMappings, labelMappings, e
+                        "type":   mapping["type"],
+                        "name":   metricPrefix + mapping['metric'],
+                        "value":  int(result["value"]),
+                        "labels": extraLabels,                   
```
```suggestion
                        "labels": extraLabels,
```

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -150,6 +161,22 @@ def _exportMetrics(self, results, metricPrefix, metricMappings, labelMappings, e
+        metrics = []
+
+        for result in results:
+            for mapping in metricMappings:                
```
```suggestion
            for mapping in metricMappings:
```

esanchezm commented on prometheus_pgbouncer_exporter/collector.py:
```diff
@@ -99,10 +99,21 @@ def collect(self):
                     {"type": "gauge", "column": "pool_size",           "metric": "database_pool_size",           "help": "Configured Pool Size Limit"},
                     {"type": "gauge", "column": "reserve_pool",        "metric": "database_reserve_pool_size",   "help": "Configured Reserve Limit"},
                     {"type": "gauge", "column": "current_connections", "metric": "database_current_connections", "help": "Database connection count"},
+                    {"type": "gauge", "column": "max_connections",     "metric": "database_max_connections",     "help": "Database maximum number of allowed connections"},                    
```
```suggestion
                    {"type": "gauge", "column": "max_connections",     "metric": "database_max_connections",     "help": "Database maximum number of allowed connections"},
```

jgduarte-stratio commented:
No worries, and thank you for your feedback
Yes I think I need to improve my IDE. I not a python dev, and I have a basic setup. 
I was also too confident about it because I was fixing al the issues the linter command was raising.

jgduarte-stratio commented:
Hi @esanchezm , I hope you're doing well. Would you have time to check everything is right now?

esanchezm commented:
@jgduarte-stratio we just released a new version 2.1.0. Please check the release notes in https://github.com/spreaker/prometheus-pgbouncer-exporter/releases

jgduarte-stratio commented:
Great news! thank you very much.
I hope we can start using it next week.

El El vie, 10 jul 2020 a las 16:11, Esteban Sanchez <
notifications@github.com> escribió:

> @jgduarte-stratio <https://github.com/jgduarte-stratio> we just released
> a new version 2.1.0. Please check the release notes in
> https://github.com/spreaker/prometheus-pgbouncer-exporter/releases
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spreaker/prometheus-pgbouncer-exporter/pull/23#issuecomment-656697395>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE6UWE4KHWISLC5YUHO6MTLR24OSPANCNFSM4NEEU5UA>
> .
>
-- 

| José Gabriel Duarte

+34 663 817 893 / Ext: 2104


Vía de las dos Castillas, 33, Ática 4, 2ª Planta

28224 Pozuelo de Alarcón, Madrid, Spain

+34 918 286 473 | www.stratio.com

<https://twitter.com/stratiobd> <https://www.linkedin.com/company/stratiobd>
<https://www.youtube.com/channel/UCoI1xy_cEsULLD8ZMiOCP1A>

------------------------------


This e-mail and the information contained in it is confidential and it is
addressed exclusively to the recipient named in the heading. If you have
received or accessed this message by error, please notify us and proceed to
erase it immediately. Any unauthorized disclosure, copying, distribution or
use of the information contained therein is prohibited by law. The personal
data provided by you or by third parties are the responsibility of STRATIO
in order to manage and maintain contacts and relationships that occur as a
result of the relationship with STRATIO. Normally, the legal basis that
legitimates this processing, will be your consent, our legitimate interest
or the need to manage a contractual relationship with you. The storage
period of your data will be determined by your relationship with us.
STRATIO's email service is outsourced with Google, which implies an
international transfer of your contact data to the United States; however,
Google provides adequate data protection guarantees, by adhering to the
Privacy Shield framework, and because of the authorization TI/00153/2017 of
the Spanish Data Protection Agency (AEPD) to transfer personal data to the
United States. For further information, or to exercise your rights of
access, rectification, erasure, objection, limitation or portability,
please write to privacy@stratio.com. Likewise, in the event that you
consider your data protection rights violated in any manner, you may file a
claim with the AEPD (www.aepd.es).
